### ISSUE TITLE
Bundlized version of ST4 and ANTLR

### ISSUE THREAD

**Issue by sakno** [antlr/stringtemplate4#88]
Could you please add maven-bundle-plugin into StringTemplate/ANTLR pom.xml and OSGify these libraries?

**Comment by cniles** [comment]
Has there been any recent consideration of this?  The change should have no impact unless an OSGi application kernel is being utilized.  If you want I'll make the changes and send in a pull request.

**Comment by parrt** [comment]
How about including the main antlr tool, which includes st 4?

**Comment by cniles** [comment]
Are you suggesting adding OSGi metadata for the monolithic antlr artifact vice adding it to the ST4 artifact, or saying that this has been done already and I should be installing the main antlr tool?

If the former then yes; that would be a good idea and I would be happy to contribute.

**Comment by parrt** [comment]
latest antlr jars are osgi compatible
T

**Comment by cniles** [comment]
Looking through the public antlr 4.5 artifacts:

The one maven artifact that has OSGi headers is org.antlr/antlr4-runtime/4.5 (antlr4-runtime-4.5jar), but it doesn't contain the ST4 classes.  The ST4 classes are in the artifact org.antlr/antlr4/4.5 but the osgi headers are not present in the jar's manifest.

Are these changes that are still in snapshot?

**Comment by cbdiesse** [comment]
Looking for antlr 4.5.x complete jar with StringTemplate-4 api exported and usable in eclipse plug-in development environment. 
The antlr.org site claim that the provided jar files are osgi ready, but it's not the case for antlr-4.5.3-complete.jar. the provided manifest.mf don't provide any package export.
On the other hand, the runtime version :  antlr-runtime-4.5.3 is osgi ready, but don't provide StringTemplate-4 packages. Meanwhile, no osgi version of StringTemplate-4 is provided.

In result, we have very hard time using ST4 and ANTLR4 in eclipse plug-in development env.
Can anyone help provide something more effective to a low profil developper like me ?  This will really help me cross the river. Thanks.

**Comment by sakno** [comment]
You can solve this problem in this way:

1. Create `antlr-bundlized` maven module in your poject
1. Define packaging `bundle` to your module
1. Specify artifact ID of ANTLR inside of `<Embed-Dependency>` tag in the configuration section of `maven-bundle-plugin` module. And don't forget to specify `scope:=compile` inside of this tag.
1. Declare dependency on ANTLR with `compile` scope (any other OSGi-friendly dependencies should be declared with `provided` scope)
1. Declare `<Embed-Transitive>` tag with value `true` in the configuration section
1. Specify `<Export-Package>` tag in the configuration section of `maven-bundle-plugin` and set `org.antlr.*` as its value

Now you can use your `antlr-bundlized` module as OSGi-friendly version of ANTLR.

**Comment by sharwell** [comment]
We've now moved all the way in this direction for the ANTLR 4 runtime. However, that library proved easier than ST4, *especially* now that we've eliminated its last external dependency. I'll look at implementing this for ST4 as well, but there's an interesting dependency chain here:

ST4 &rarr; ANTLR 3 &rarr; ST3 &rarr; ANTLR 2

The real reason this is so interesting is twofold:

1. ST4 doesn't use any features of ANTLR 3 which depend on ST3. However, ANTLR 3 as an independent bundle *would* need to declare the ST3 dependency.
2. The shaded ST4 bundle wouldn't have any external dependencies, but maven-bundle-plugin runs before maven-shade-plugin so the manifest for the complete jar still lists imports for the ANTLR 3 runtime, even though they aren't actually required. I found [this answer](http://stackoverflow.com/a/18601409/138304) on Stack Overflow which suggests solutions to this exist, but I haven't followed this investigation to the end yet.

**Comment by felixdo** [comment]
I would like to play with this, but when I run mvn clean verify, the resulting jar does not contain the 'st4hidden' package that is in the jar I can download from the st website. What am I missing?

**Comment by parrt** [comment]
@felixdo Does the resulting jar not operate as you would expect?

**Comment by felixdo** [comment]
No. I would like to produce a standalone OSGI bundle with no external dependencies for st4, which afaiu is the goal of this issue, and which does not exist atm. I would therefore need to reproduce include the antlr runtime just as it is included in ST-4.0.8.jar that I get from the st4 website. mvn clean verify does however not reproduce that jar, it misses the antlr runtime.

**Comment by parrt** [comment]
oh, well i guess we need to use the mvn shade thing to build a combined lib.

**Comment by felixdo** [comment]
yes that's what I meant. I thought the pom here already included that part.

**Comment by felixdo** [comment]
Maybe not even needed to include antlr, since that is already available in eclipse orbit https://dev.eclipse.org/ipzilla/show_bug.cgi?id=4865, so if we produce a bundle with a dependency on that, it should work. or not, if i still don't understand the dependency chain here... :)
